channeldb/migration26/iterator.go

```diff
@@ -0,0 +1,257 @@
+package migration26
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	mig25 "github.com/lightningnetwork/lnd/channeldb/migration25"
+
+	"github.com/lightningnetwork/lnd/kvdb"
+)
+
+var (
+	// openChanBucket stores all the currently open channels. This bucket
+	// has a second, nested bucket which is keyed by a node's ID. Within
+	// that node ID bucket, all attributes required to track, update, and
+	// close a channel are stored.
+	openChannelBucket = []byte("open-chan-bucket")
+
+	// errExit is returned when the callback function used in iterator
+	// needs to exit the iteration.
+	errExit = errors.New("exit condition met")
+)
+
+// updateLocator defines a locator that can be used to find the next record to
+// be migrated. This is useful when an interrupted migration that leads to a
+// mixed revocation log formats saved in our database, we can then restart the
+// migration using the locator to continue migrating the rest.
+type updateLocator struct {
+	// nodePub, chainHash and fundingOutpoint are used to locate the
+	// channel bucket.
+	nodePub         []byte
+	chainHash       []byte
+	fundingOutpoint []byte
+
+	// nextHeight is used to locate the next old revocation log to be
+	// migrated. A nil value means we've finished the migration.
+	nextHeight []byte
+}
+
+// fetchChanBucket is a helper function that returns the bucket where a
+// channel's data resides in given: the public key for the node, the outpoint,
+// and the chainhash that the channel resides on.
+func (ul *updateLocator) locateChanBucket(rootBucket kvdb.RwBucket) (
+	kvdb.RwBucket, error) {
+
+	// Within this top level bucket, fetch the bucket dedicated to storing
+	// open channel data specific to the remote node.
+	nodeChanBucket := rootBucket.NestedReadWriteBucket(ul.nodePub)
+	if nodeChanBucket == nil {
+		return nil, mig25.ErrNoActiveChannels
+	}
+
+	// We'll then recurse down an additional layer in order to fetch the
+	// bucket for this particular chain.
+	chainBucket := nodeChanBucket.NestedReadWriteBucket(ul.chainHash)
+	if chainBucket == nil {
+		return nil, mig25.ErrNoActiveChannels
+	}
+
+	// With the bucket for the node and chain fetched, we can now go down
+	// another level, for this channel itself.
+	chanBucket := chainBucket.NestedReadWriteBucket(ul.fundingOutpoint)
+	if chanBucket == nil {
+		return nil, mig25.ErrChannelNotFound
+	}
+
+	return chanBucket, nil
+}
+
+// findNextMigrateHeight finds the next commit height that's not migrated. It
```

Comment:
Very cool, so this is intended to enable us to not have to commit the entire migration in a single transaction and to support incremental restart?